# this should be the only path that you need to modify
BASE_PATH = "."

N_THREADS = 40

# annotation stuff
IDX = BASE_PATH + "/index"

ANNO = BASE_PATH + "/annotation"
ANNO_PREFIX = "Homo_sapiens.GRCh38.80"
ANNO_GTF = "{0}/{1}.gtf".format( ANNO, ANNO_PREFIX )
ANNO_FA = "{0}/{1}.fa".format( ANNO, ANNO_PREFIX )
ANNO_BWT = "{0}/{1}".format( IDX, ANNO_PREFIX )
ANNO_TOPHAT = "{0}/{1}_tophat".format(IDX, ANNO_PREFIX)
ANNO_RSEM = "{0}/{1}_rsem/ref".format( IDX, ANNO_PREFIX )

GENOME_NAME = "GRCh38_80"
GENOME_BASE_PATH = "{0}/{1}".format(BASE_PATH, "genome")
GENOME_FA = GENOME_BASE_PATH + "/Homo_sapiens.GRCh38.dna.primary_assembly.fa"
GENOME_BWT = GENOME_BASE_PATH + "/Homo_sapiens.GRCh38.dna.primary_assembly"

ERCC_ANNO_FA = "{0}/{1}_ercc.fa".format(ANNO, ANNO_PREFIX)
ERCC_ANNO_GTF = "{0}/{1}_ercc.gtf".format(ANNO, ANNO_PREFIX)

ERCC_GENOME_FA = "{0}/{1}_ercc.fa".format(IDX, GENOME_NAME)

KAL_IDX = "{0}/{1}.kidx".format( IDX, ANNO_PREFIX )
SAILFISH_IDX = "{0}/{1}_sailfish_idx".format( IDX, ANNO_PREFIX )

HISAT_IDX = "{0}/hisat_{1}/{1}".format(IDX, GENOME_NAME)
HISAT_SPLICESITES = "{0}/{1}_splicesites.txt".format(ANNO, ANNO_PREFIX)

ERCC_FA = '{0}/ERCC.fa'.format(IDX)
ERCC_KAL_IDX = "{0}/{1}_ercc.kidx".format(IDX, ANNO_PREFIX)
ERCC_SAILFISH_IDX = "{0}/{1}_ercc_sailfish_idx".format(IDX, ANNO_PREFIX)
ERCC_RSEM_DIR = "{0}/{1}_ercc_rsem".format(IDX, ANNO_PREFIX)
ERCC_ANNO_BWT = "{0}/{1}_ercc".format(IDX, ANNO_PREFIX)
ERCC_GENOME_BWT = "{0}/{1}_ercc".format(IDX, GENOME_NAME)
ERCC_EMSAR_100_IDX = "{0}/{1}_ercc_emsar_100.rsh".format(IDX, ANNO_PREFIX)
ERCC_EMSAR_101_IDX = "{0}/{1}_ercc_emsar_101.rsh".format(IDX, ANNO_PREFIX)

SOFTWARE_PRE = BASE_PATH + '/software'
KALLISTO = '{0}/kallisto-0.42.3/kallisto'.format(SOFTWARE_PRE)
SAILFISH = 'LD_LIBRARY_PATH={0}/{1}/lib {0}/{1}/bin/sailfish'.format(SOFTWARE_PRE, "Sailfish-0.6.3-Linux_x86-64")
FLUX = '{0}/flux-capacitor-1.6.1/bin/flux-capacitor'.format(SOFTWARE_PRE)
EMSAR = '{0}/emsar/emsar'.format(SOFTWARE_PRE)
EMSAR_BUILD = '{0}/emsar/emsar-build'.format(SOFTWARE_PRE)
HISAT_DIR = '{0}/hisat-0.1.6-beta'.format(SOFTWARE_PRE)
HISAT = '{0}/hisat'.format(HISAT_DIR)

# initialization
def create_files_and_directories():
    import os
    # Create directories
    os.makedirs(IDX, exist_ok=True)
    os.makedirs(ANNO, exist_ok=True)
    os.makedirs(GENOME_BASE_PATH, exist_ok=True)
    # Create files
    open(GENOME_FA, 'w').close()
    open(ERCC_FA, 'w').close()
    open(ANNO_FA, 'w').close()
    open(ANNO_GTF, 'w').close()
    

create_files_and_directories()

# functions

def source_r(base, fname):
    return 'Rscript --vanilla --default-packages=methods,stats,utils -e \'setwd("{0}")\' -e \'source("{1}")\''.format(base, fname)

def source_rmd(base, file_name):
    return 'Rscript --vanilla --default-packages=methods,stats,utils,knitr -e \'setwd("{0}")\' -e \'rmarkdown::render("{1}")\''.format(base, file_name)

rule all:
    input:
        ERCC_ANNO_GTF,
        ERCC_KAL_IDX,
        ERCC_SAILFISH_IDX,
        ERCC_RSEM_DIR + '/ref.grp',
        ERCC_RSEM_DIR + '/ref.transcripts.1.bt2',
        ERCC_GENOME_BWT + '.1.bt2',
        ERCC_ANNO_BWT + '.1.bt2',
        ERCC_EMSAR_100_IDX,
        ERCC_EMSAR_101_IDX,
        expand("simulations/NA12716_7/NA12716_7_{end}.fastq.gz", end = [1,2]),
        expand("personalized_simulation/NA12716_7/NA12716_7_{end}.fastq.gz", end = [1,2]),
        "annotation/Homo_sapiens.GRCh38.80.gtf",
        "annotation/Homo_sapiens.GRCh38.80.fa",
        "genome/Homo_sapiens.GRCh38.dna.primary_assembly.fa",
        "annotation/NA12716.fa",
        "personalized_simulation/NA12716_7/NA12716_7.hap"


rule ercc_gtf:
     input:
        ERCC_FA,
        ANNO_GTF
     output:
        ERCC_ANNO_GTF
     run:
        ercc = open(ERCC_FA).readlines()
        num_ercc = int(len(ercc)/2)
        shell("cp {ANNO_GTF} {ERCC_ANNO_GTF}")
        out = open(ERCC_ANNO_GTF, "a")
        for i in range(num_ercc):
            ercc_name = ercc[2*i].split()[0][1:]
            ercc_len = len(ercc[2*i + 1]) - 1
            out.write("{0}\tERCC\tgene\t1\t{1}\t.\t+\t.\tgene_id \"{0}\"; gene_name \"{0}\";\n".format(ercc_name, ercc_len))
            out.write("{0}\tERCC\ttranscript\t1\t{1}\t.\t+\t.\tgene_id \"{0}\"; transcript_id \"{0}\"; gene_name \"{0}\";\n".format(ercc_name, ercc_len))
            out.write("{0}\tERCC\texon\t1\t{1}\t.\t+\t.\tgene_id \"{0}\"; transcript_id \"{0}\"; gene_name \"{0}\";\n".format(ercc_name, ercc_len))
        out.close()


rule bwt2_genome:
    input:
        GENOME_FA
    output:
        expand(GENOME_BWT + ".{i}.bt2", i = range(1, 5)),
        expand(GENOME_BWT + ".rev.{i}.bt2", i = range(1, 3))
    threads: 1
    shell:
        'bowtie2-build '
        '--offrate 1 '
        '--seed 37 '
        '{input} {GENOME_BWT}'


rule bwt2_genome_ercc:
    input:
        ERCC_GENOME_FA
    output:
        expand(ERCC_GENOME_BWT + ".{i}.bt2", i = range(1, 5)),
        expand(ERCC_GENOME_BWT + ".rev.{i}.bt2", i = range(1, 3))
    threads: 1
    shell:
        'bowtie2-build '
        '--offrate 1 '
        '--seed 37 '
        '{input} {ERCC_GENOME_BWT}'


rule bwt2_anno_ercc:
    input:
        ERCC_ANNO_FA
    output:
        expand(ERCC_ANNO_BWT + ".{i}.bt2", i = range(1, 5)),
        expand(ERCC_ANNO_BWT + ".rev.{i}.bt2", i = range(1, 3))
    threads: 1
    shell:
        'bowtie2-build '
        '--offrate 1 '
        '--seed 37 '
        '{input} {ERCC_ANNO_BWT}'


rule kal_ercc_idx:
    input:
        ERCC_ANNO_FA
    output:
        ERCC_KAL_IDX
    shell:
        KALLISTO + ' index -i {output} {input}'


rule merge_anno_ercc:
    input:
        ANNO_FA,
        ERCC_FA
    output:
        ERCC_ANNO_FA
    shell:
        'cat {input[1]} {input[0]} > {output}'

rule merge_genome_ercc:
    input:
        GENOME_FA,
        ERCC_FA
    output:
        ERCC_GENOME_FA
    shell:
        'cat {input[1]} {input[0]} > {output}'

rule sail_ercc_idx:
    input:
        ERCC_ANNO_FA
    output:
        ERCC_SAILFISH_IDX
    threads: N_THREADS
    shell:
        SAILFISH + ' index '
        '-t ' + ERCC_ANNO_FA + ' '
        '-o {output} '
        '-k 21'


rule rsem_prepare_ercc:
    input:
        ERCC_ANNO_FA
    output:
        "{0}/ref.grp".format(ERCC_RSEM_DIR)
    run:
        shell('mkdir -p {0}'.format(ERCC_RSEM_DIR))
        shell('rsem-prepare-reference {0} {1}/ref'.format(ERCC_ANNO_FA, ERCC_RSEM_DIR))


rule rsem_bwt2_idx:
    input:
        "{0}/ref.grp".format(ERCC_RSEM_DIR)
    output:
        "{0}/ref.transcripts.1.bt2".format(ERCC_RSEM_DIR),
        "{0}/ref.transcripts.2.bt2".format(ERCC_RSEM_DIR),
        "{0}/ref.transcripts.3.bt2".format(ERCC_RSEM_DIR),
        "{0}/ref.transcripts.4.bt2".format(ERCC_RSEM_DIR),
        "{0}/ref.transcripts.rev.1.bt2".format(ERCC_RSEM_DIR),
        "{0}/ref.transcripts.rev.2.bt2".format(ERCC_RSEM_DIR)
    shell:
        'bowtie2-build '
        '--seed 42 '
        '--offrate 1 '
        '{ERCC_RSEM_DIR}/ref.transcripts.fa '
        '{ERCC_RSEM_DIR}/ref.transcripts'


rule emsar_ercc_index_100:
    input:
        ERCC_ANNO_FA
    output:
        ERCC_EMSAR_100_IDX
    threads:
        N_THREADS
    run:
        SOFTWARE_PRE + '/emsar/emsar-build -P -p {threads} {ERCC_ANNO_FA} 100 {IDX} {ANNO_PREF}_ercc_emsar_100'


rule emsar_ercc_index_101:
    input:
        ERCC_ANNO_FA
    output:
        ERCC_EMSAR_101_IDX
    threads:
        N_THREADS
    run:
        SOFTWARE_PRE + '/emsar/emsar-build -P -p {threads} {ERCC_ANNO_FA} 101 {IDX} {ANNO_PREF}_ercc_emsar_101'


rule get_geuvadis:
     output:
        "simulations/NA12716_7/NA12716_7_{end}.fastq.gz"
     shell:
        "wget -O simulations/NA12716_7/NA12716_7_{wildcards.end}.fastq.gz ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR188/ERR188140/ERR188140_{wildcards.end}.fastq.gz; "
        "ln -s simulations/NA12716_7/NA12716_7_{wildcards.end}.fastq.gz personalized_simulation/NA12716_7/NA12716_7_{wildcards.end}.fastq.gz"


rule make_symlinks:
     input:
        "simulations/NA12716_7/NA12716_7_{end}.fastq.gz"
     output:
        "personalized_simulation/NA12716_7/NA12716_7_{end}.fastq.gz"
     shell:
        "ln -s ../../simulations/NA12716_7/NA12716_7_{wildcards.end}.fastq.gz personalized_simulation/NA12716_7/NA12716_7_{wildcards.end}.fastq.gz"


rule get_genome:
     output:
        "genome/Homo_sapiens.GRCh38.dna.primary_assembly.fa"
     shell:
        "cd genome; "
        "wget ftp://ftp.ensembl.org/pub/release-80/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz; "
        "gunzip Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz"


rule get_anno:
     output:
        "annotation/Homo_sapiens.GRCh38.80.gtf",
        "annotation/Homo_sapiens.GRCh38.80.fa"
     shell:
        "cd annotation; "
        "wget ftp://ftp.ensembl.org/pub/release-80/gtf/homo_sapiens/Homo_sapiens.GRCh38.80.gtf.gz; "
        "gunzip Homo_sapiens.GRCh38.80.gtf.gz; "
        "gffread Homo_sapiens.GRCh38.80.gtf -g ../genome/Homo_sapiens.GRCh38.dna.primary_assembly.fa -w Homo_sapiens.GRCh38.80.fa; "


rule get_personal:
     output:
        "annotation/NA12716.fa",
        "personalized_simulation/NA12716_7/NA12716_7.hap"
     shell:
        "wget http://lmcb.math.berkeley.edu/kallisto/NA12716.fa; "
        "cd ../personalized_simulation/NA12716_7/; "
        "wget http://lmcb.math.berkeley.edu/kallisto/NA12716.hap; " 
        
